### PR TITLE
Wrap microcluster app and API endpoints

### PR DESCRIPTION
**Pull request by neoaggelos**
### Summary

Capture all required state for the microcluster and the API endpoint implementation in structs.

### Changes

- `*app.App` now wraps the microcluster app and the snap interface
- `*app.App` now implements `api.Provider`
- `*api.Endpoints` now wraps the required provider (to have access to the snap and microcluster app), and implements the API endpoints
- Do away with the `wrapHandlerWithMicrocluster` closure, and retrieving the snap through the server context.

### Notes

This will allow us to add more configs to the application, and works towards making it easier to write tests for our API endpoints

### CODE REVIEW

**Comment by neoaggelos** [comment]
Blocked by #282 and #279, as there are conflicts

**Review by berkayoz — APPROVED** [review]
Nice work, LGTM.

File: src/k8s/pkg/k8sd/api/endpoints.go [hunk 1]
```diff
@@ -2,55 +2,63 @@
...
+}
+
+// New creates a new Endpoints instance.
+func New(provider Provider) *Endpoints {
+	return &Endpoints{provider: provider}
+}
+
 // Endpoints returns the list of endpoints for a given microcluster app.
-func Endpoints(app *microcluster.MicroCluster) []rest.Endpoint {
+func (e *Endpoints) Endpoints() []rest.Endpoint {
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/api/endpoints.go:18]: nit: `Endpoints.Endpoints()` reads a bit weird. Maybe call this `API`?

File: src/k8s/pkg/k8sd/app/app.go [hunk 2]
```diff
@@ -55,10 +55,10 @@ func New(ctx context.Context, cfg Config) (*App, error) {
 func (a *App) Run(customHooks *config.Hooks) error {
 	// TODO: consider improving API for overriding hooks.
```
Comment (bschimke95) [inline src/k8s/pkg/k8sd/app/app.go:56]: nit: is this still relevant? Seems outdated.

**Review by bschimke95 — APPROVED** [review]
Looks great overall. Only minor comments

File: src/k8s/pkg/k8sd/app/app.go (hunk 2 above)
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/app/app.go:56]: yeah, we now pass a `*config.Hooks` and selectively apply some of the non-nil hooks. which works _ok_, as we do hange some hooks in a couple of tests.

but this is not too clean, i'd keep the TODO around for the future

File: src/k8s/pkg/k8sd/api/endpoints.go (hunk 1 above)
Comment (neoaggelos) [reply src/k8s/pkg/k8sd/api/endpoints.go:18]: Usage is `api.New(app, snap).Endpoints()`, should we maybe rename the type to `*api.API`?

**Pull request merged**